// @adminjs/design-system/build/molecules/rich-text-editor/rich-text-editor.styled.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuBarWrapper = exports.EditorWrapper = void 0;
const styled_components_1 = __importDefault(require("styled-components"));
const box_1 = __importDefault(require("../../atoms/box/box"));
const input_1 = require("../../atoms/input");
const text_1 = require("../../atoms/text");
const utils_1 = require("../../utils");
exports.EditorWrapper = styled_components_1.default(text_1.Text) `
  position: relative;
  ${input_1.InputCSS}

  .ProseMirror-focused {
    outline: none;
  }

  .characterCount {
    position: absolute;
    bottom: ${utils_1.themeGet('space', 'md')};
    right: ${utils_1.themeGet('space', 'lg')};
    color: ${utils_1.themeGet('colors', 'grey40')};
  }
`;
exports.EditorWrapper.defaultProps = {
    px: 'xl',
    py: 'xl',
    className: utils_1.cssClass('EditorWrapper'),
};
exports.MenuBarWrapper = styled_components_1.default(box_1.default) `
  border: 1px solid ${utils_1.themeGet('colors', 'inputBorder')};
  border-bottom: none;

  & .${utils_1.cssClass('Icon')}:hover {
    cursor: pointer;
  }

  & .${utils_1.cssClass('Icon')} svg {
    fill: ${utils_1.themeGet('colors', 'grey100')};
  }

  & .active .${utils_1.cssClass('Icon')} svg {
    fill: ${utils_1.themeGet('colors', 'primary100')};
  }
`;
exports.MenuBarWrapper.defaultProps = {
    px: 'md',
    py: 'md',
    className: utils_1.cssClass('MenuBarWrapper'),
};
